Explainer for PA per-participant metrics (#1272)
WICG/turtledove: FLEDGE_extended_PA_reporting.md

matter whether a bid wins the auction or not. The browser will always trigger reporting for the
`always` contributions.

## Per-participant metrics

More recent versions of Chrome offer some additional features, focused on measuring resource
usage and whole-auction metrics rather than those specific to a particular function execution
or the winning bid.

> First, the new `reserved.once` event-type is a special value that, for each sub-auction, selects a

Reviewer: Glad the contributions limit is increasing to 100.
https://chromestatus.com/feature/5114676393017344

> First, the new `reserved.once` event-type is a special value that, for each sub-auction, selects a

Reviewer: consistency nit:  sub-auction -> sub(auction)

Author: Done.

random invocation of `generateBid()` and of `scoreAd()` independently, and reports private

> aggregation contributions with that event only from those executions. (In case of a multi-seller

Reviewer: multi-seller -> component

Author: I went with a longer "an auction with component auctions" to avoid potential confusion..

The event may not be used in `reportWin()` or `reportResult()`; since those already run at most
once per auction level, `reserved.always` may be used instead.

This feature is intended for reporting overall per-participant metrics only once rather than for
every interest group participating. A number of new `baseValues` representing such values are
available and described below, but it can also be useful with per-IG metrics which are not expected to
vary much like `signals-fetch-time`, to sample them.

While usage of `reserved.once` will be ignored by older versions, newly added `baseValues` will not
be, so the calls to `contributeToHistogramOnEvent()` should be individually wrapped in `try/catch`.
That is also encouraged in general since `contributeToHistogramOnEvent()` is specified to throw
on permission policy violations.

Users using this are strongly encouraged to report their metrics during the beginning of their
scripts, since if the script hits a per script time out before asking to report them nothing will
get sent, which can result in inaccuracy, especially for `percent-scripts-timeout`.

### Per-participant base values.

The newly added base values are as following:

> * `participating-ig-count`: number of interest groups that got a chance to participate in this

Reviewer: Noting that `participating-ig-count` slightly different than `interestGroupCount` per-buyer measurement available to buyers.

>   (sub)auction, i.e. they had registered ads, and if no timeouts were to hit, their `generateBid()`

Author: It's supposed to be a hypothetical?

>   (sub)auction, i.e. they had registered ads, and if no timeouts were to hit, their `generateBid()`

Reviewer: Ah, I see. Perhaps...?

i.e. they had registered ads and their `generateBid()` was queued/scheduled to run.

(timeouts aren't the only thing that could deep six their participation. the auction could be aborted, &c. )

>   (sub)auction, i.e. they had registered ads, and if no timeouts were to hit, their `generateBid()`

Author: I don't really want to talk about queueing/scheduling since it's a bit of an implementation detail --- e.g. in our implementation generateBid isn't actually scheduled to run until all of relevant promises resolve. 
Anyway, tried to rephrase it to be more concrete.

(Also aborts presently completely drop PA ... which I suspect we will need to change...)

  would have been run.
* `average-code-fetch-time`: average time of all code fetches (including JavaScript and WASM) used
  for all invocations of particular function type for given (sub)auction.

> * `percent-scripts-timeout`: percentage of script executions of this kind that hit the per-script

Reviewer: To verify, for a seller this would be the percentage of `scoreAd` executions that hit `sellerTimeout` || 50 and `reportResult` executions that hit `reportingTimeout`  || 50.

And for a buyer the percentage of `generateBid` executions that hit a `perBuyerTimeout` || 50 and `reportWin` executions that hit `reportingTimeout`  || 50.

Author: I am not sure what exactly you mean by || 50, but it's not combining scoreAd and reportResult and generateBid/reportWin.

scoreAd would get its own time outs, and reportResult its own (which is always 0 or 100). So you would want to report it as "reserved.once" to some range of buckets for scoreAd and as "reserved.always" to another range of buckets for  reportResult. (And if you start seeing some timeouts, you might decided that you need bigger sellerTimeout or bigger reportingTimeout respectively).

The big table below is supposed to be explaining this...

Reviewer: >I am not sure what exactly you mean by || 50

As I read the explainer/impl guide, 50 is the script-level default absent a relevant timeout having been configured.

Reviewer: >(And if you start seeing some timeouts, you might decided that you need bigger sellerTimeout or bigger reportingTimeout respectively).

Makes sense for a _seller_ observing their script timeouts as they control the `sellerTimeout` knob. 

But a buyer seeing timeouts cannot take action since the seller(s) set `reportingTimeout`.

Author: Oh, yeah, 50 being default is true. But it's separate measurements between scoreAd and reportResult (and likewise for buyer) --- exactly so people can check whether reportingTimeout and sellerTimeout/perBuyerTimeout are set correctly independently.

>   timeout.

Reviewer: >percentage of script executions of this kind that hit the per-script timeout

This _of this kind_ reading would be accessible in reporting functions via the suggested use of `reserved.always`.

Author: Not sure what you mean here?

Reviewer: Poor wording as I wrapped my head around values not strictly associated with `reserved.once`.

>`percent-scripts-timeout`: percentage of script executions of this kind that hit the per-script timeout.

My comment was recognizing that one must use `reserved.always` in a reporting function. Since this baseValue actually measures invocations of the function in which the histogram contribution is reported, might the wording be "of this function?"

>while percent-scripts-timeout in `reportWin()` is either 0 or 100 dependent on whether the reporting function's execution timed out or not (assuming reporting is done early enough to happen if it did)

This applies to `scoreAd()` also, right? Maybe "in reporting functions?"

<br/>

>`average-code-fetch-time`: average time of all code fetches (including JavaScript and WASM) used for all invocations of particular function type for given (sub)auction.

I read this as _different measurement_ by function type (bid/score versus reporting), but until Chrome introduces a separate reporting logic url ([#679](https://github.com/WICG/turtledove/issues/679#issuecomment-1703973736)) there's no differentiation, right? Because the fetches are part of the executor setup. 

To verify, whether using `.once` or `.always` (in reportResult), sellers will receive the same value - their scoringLogicURL's code fetch time.

Buyers would also receive the same value regardless of where the contribution is reported.  
A buyer with IGs using only biddingLogicURLs would receive the average code fetch across IGs that made it that far. 
A buyer with IGs using WASM as well would receive the average code fetch irrespective of type.

Author: So reporting code fetch time being the same as for generateBid/scoreAd is not not necessarily true, since we have a process limit, and so the process may actually need to be started fresh for reporting (and fetch fresh... hopefully from cache). That's actually the common case for reportWin (basically unless a concurrent auction keeps the process alive).

Reviewer: Oh, yes, right. You don't hold onto that [limited resource](https://github.com/WICG/turtledove/blob/main/PA_implementation_overview.md#:~:text=There%20are%20separate,to%20resource%20constraints.
). 

>There are separate Chrome-wide limits on the number of buyer processes (10) and seller processes (3).  These numbers are not optimized, and subject to change.  On mobile, Chrome will generally reuse the single renderer process with one thread for bidders and one for sellers, instead of using more secure, isolated processes, due to resource constraints.

Verifying that "on mobile" equates to 
```
Sec-CH-UA-Mobile: ?1
```

Author: No, it equates to "Android" (so telling devtools to emulate an iPhone on your laptop won't change that... not would "request desktop site" on a phone). I suppose it should probably just say that.

* `percent-igs-cumulative-timeout`: percentage of interest groups from this buyer that did not get

>   to participate in this (sub)auction due to the per-buyer cumulative timeout

Reviewer: The implementation guide says
>Groups with matching joining origins and priority are randomly shuffled.

If a buyer is not using any priority and has three IGs joined each joined on a separate origin these are processed in whatever order they were returned form SQLlite?

Author: I think they'll end up ordered by join origin, not sure in which direction. Hmm, good reminder that `participating-ig-count` is after the filtering.

Reviewer: Duh, yes, it does say ordered by join origin. 🤦

>   (`participating-ig-count` is the quotient here).

Author: Woops, yep. Thank you.

* `cumulative-buyer-time`: total time spent for buyer's computation, in milliseconds; this is what

>   would normally be  compared against the per-buyer cumulative timeout. If the timeout is not hit,

Reviewer: nit: be  compared   (extra space)

Author: Done

>   the value is capped at the per-buyer cumulative timeout, if it's hit, the value will be the

Reviewer: >If the timeout is not hit, the value is capped at the per-buyer cumulative timeout, if it's hit, the value will be the timeout + 1.

The seller has configured `perBuyerCumulativeTimeout` of 500. This value wuill either be `500` or `501`?
And if no value was configured?

Author: Good point, I should clarify that (thanks!). To answer meanwhile:
If no value was configured you'll just get a 0.
If the value was set to 500, you are not limited to just 500 or 501 --- like if it took 400ms you'll get 400ish ms. 

The "capped" bit is that there is some sloppiness in measurements, so it's possible that nothing got timed out but  it actually got measured as 501ms. And similarly the timeout may end up measured at 503ms or whatever, so it's trying to normalize these so you can tell them apart. (Though maybe I should use something more than +1 for easier bucketing?)

Author: Rephrased a bit

  timeout + 1.
* `percent-regular-ig-count-quota-used`,`percent-negative-ig-count-quota-used`,
  `percent-ig-storage-quota-used`: percentage of the database quote used by the buyer for
  regular interest group count, negative targeting interest group count, and overall byte usage
  respectively. This is capped at 110 since the quotas may not be enforced immediately (and actual
  usage in that case may be bigger than 110%).

Note that these metrics are measured only for some kinds of worklet executions &mdash; some are
only relevant for bidders, and get 0 in the seller functions. In case of reporting functions,
they sometimes repeat what was available in the corresponding `generateBid()` or `scoreAd()`,
and sometimes get their own measurement. This is shown below:

| `baseValue` name | In `generateBid() ` | In `reportWin()` | In `scoreAd()` | In `reportResult()` |
| ---------------- | ------------------- | ---------------- | -------------- | ------------------- |
| `participating-ig-count`  | Measured | From `generateBid()`  | 0 | 0 |

> | `average-code-fetch-time` | Measured | Measured | Measured | Measured |

Reviewer: Readability suggestion: move the common values, `average-code-fetch-time` & `percent-scripts-timeout` to the top/bottom of the list.

Author: Done

matter whether a bid wins the auction or not. The browser will always trigger reporting for the
`always` contributions.

## Per-participant metrics

> More recent versions of Chrome offer some additional features, focused on measuring resource

Reviewer: M130 intends to enable Bidding & Auction Services by default. Does the new Per-Participant capabilities for now apply only to "classic" FLEDGE on-device auctions?

Author: Practically, yeah, there is no support for B&A. It would need to be added to the B&A server (when it sends over PrivateAggregation stuff it just sends over the computed bucket and value numbers, no baseValues at this point).

... But I think it won't need any Chrome-side changes.

matter whether a bid wins the auction or not. The browser will always trigger reporting for the
`always` contributions.

## Per-participant metrics

More recent versions of Chrome offer some additional features, focused on measuring resource
usage and whole-auction metrics rather than those specific to a particular function execution
or the winning bid.

First, the new `reserved.once` event-type is a special value that, for each (sub)auction, selects a
random invocation of `generateBid()` and of `scoreAd()` independently, and reports private

> aggregation contributions with that event only from those executions. (In case of an auction

Author: Done

with component auctions, the top-level auction will have a single `scoreAd()` invocation selected as
well).

The event may not be used in `reportWin()` or `reportResult()`; since those already run at most
once per auction level, `reserved.always` may be used instead.

This feature is intended for reporting overall per-participant metrics only once rather than for
every interest group participating. A number of new `baseValues` representing such values are

> available and described below, but it can also be useful with per-IG metrics which are not expected to

Author: Done (and tweaked the sentence a bit, it was missing a comma)

> vary much like `signals-fetch-time`, to sample them.

Author: Done

While usage of `reserved.once` will be ignored by older versions, newly added `baseValues` will not
be, so the calls to `contributeToHistogramOnEvent()` should be individually wrapped in `try/catch`.
That is also encouraged in general since `contributeToHistogramOnEvent()` is specified to throw
on permission policy violations.

Users using this are strongly encouraged to report their metrics during the beginning of their

> scripts, since if the script hits a per script time out before asking to report them nothing will

Author: Done

get sent, which can result in inaccuracy, especially for `percent-scripts-timeout`.

### Per-participant base values.

The newly added base values are as following:
* `participating-ig-count`: number of interest groups that got a chance to participate in this
  (sub)auction, i.e. they had registered ads, did not have unsatisfied capabilities, and were not
  filtered based on priority. Interest groups included in this might not actually get to bid if the

>   cumulative timeout expires, or the script fails to load, etc, but they would have if nothing went
>   wrong.

Reviewer 2: ```suggestion
  cumulative timeout expires, or the script fails to load, etc, but they would have otherwise.
```
I don't want to stigmatize hitting the cumulative timeout as "wrong" as it may be working-as-intended.

Author: Done

* `average-code-fetch-time`: average time of all code fetches (including JavaScript and WASM) used
  for all invocations of particular function type for given (sub)auction.

> * `percent-scripts-timeout`: percentage of script executions of this kind that hit the per-script
>   timeout.

Reviewer 2: What is "per-script timeout"?  is that `perBuyerTimeouts` but not `perBuyerCumulativeTimeouts`? if so, I think we should list all of them like:
```suggestion
* `percent-scripts-timeout`: percentage of script executions of this kind that hit `perBuyerTimeouts`,
  `sellerTimeout`, or `reportingTimeout`.
```

Author: Done

* `percent-igs-cumulative-timeout`: percentage of interest groups from this buyer that did not get
  to participate in this (sub)auction due to the per-buyer cumulative timeout

>   (`participating-ig-count` is the denominator here).

Reviewer 2: The denominator comment might confuse folks into thinking they need to divide by it.  This might clarify:
```suggestion
  out of the interest groups that could have participated (see `participating-ig-count`
  above for description of participation qualifications).
```

Author: Done, also added a similar note to above.

> * `cumulative-buyer-time`: total time spent for buyer's computation, in milliseconds; this is what

Reviewer 2: I worry that "computation" may not be understood to include things like downloading.
```suggestion
* `cumulative-buyer-time`: total time spent for buyer's bid generation (downloading resources,
  starting processes, and all generate bid calls), in milliseconds; this is what
```

Author: Done, but starting processes is explicitly not included.

Reviewer 2: Hmm `perBuyerCumulativeTimeouts` explainer text says it includes starting processes.

Author: Oh. I am wrong (and bidder_process_received_ is confusingly named). We start the timer after the process is /assigned/ and not after it actually finished starting, so it is included... depending on your definition of "starting processes" anyway.

  would normally be compared against the per-buyer cumulative timeout (which must be set for this
  to be non-zero). If the timeout is not hit, the value will be how long the buyer actually took,
  capped by the per-buyer cumulative timeout, if the timeout is hit, the reported value will be the

>   timeout + 1000.

Reviewer 2: Might want to explain how the "+ 1000" part makes this more useful.

Author: Done?

* `percent-regular-ig-count-quota-used`,`percent-negative-ig-count-quota-used`,
  `percent-ig-storage-quota-used`: percentage of the database quota used by the buyer for
  regular interest group count, negative targeting interest group count, and overall byte usage
  respectively. This is capped at 110 since the quotas may not be enforced immediately (and actual
  usage in that case may be bigger than 110%).
* `regular-igs-count`, `negative-igs-count`, `ig-storage-used`: the raw counts for the buyer's

>   number of regular interest group, negative targeting interest groups, and overall byte usage,

Reviewer 2: Hmm will `ig-storage-used` hit the L1 local contribution budget cap?  If so, I guess users can scale this down, but perhaps we should mention that here.

Author: Done?
